### PR TITLE
expose local-hash as a config so developers can optionally enable fas…

### PR DESCRIPTION
**Pull request by imzhenyu**
…t rpc lookup optimization, to fix issue #279

### CODE REVIEW

**Comment by imzhenyu** [comment]
@shengofsun @qinzuoyan please help review

File: src/core/core/service_engine.cpp [hunk 1]
```diff
@@ -415,6 +415,16 @@ void service_engine::init_before_toollets(const service_spec& spec)
         spec.perf_counter_factory_name.c_str(), ::dsn::PROVIDER_TYPE_MAIN
         )
         );
+
+    // init common for all per-node providers
+    message_ex::s_local_hash = (uint32_t)dsn_config_get_value_uint64(
+        "core",
+        "local-hash",
```
Comment (qinzuoyan) [inline src/core/core/service_engine.cpp]: should be "local_hash"? all config keys are separated by '_'.

File: src/core/core/rpc_engine.cpp [hunk 2]
```diff
@@ -456,9 +456,10 @@ namespace dsn {
     rpc_request_task* rpc_server_dispatcher::on_request(message_ex* msg, service_node* node)
     {
         rpc_handler_info* handler = nullptr;
-        if (msg->header->rpc_name_fast.local_binary_hash == message_ex::s_local_binary_hash)
+        auto binary_hash = msg->header->rpc_name_fast.local_hash;
+        if (binary_hash == message_ex::s_local_hash && binary_hash != 0)
```
Comment (qinzuoyan) [inline src/core/core/rpc_engine.cpp:460]: check "binary_hash != 0" firstly, because at most time binary_hash may be 0

Comment (imzhenyu) [inline src/core/core/rpc_engine.cpp:460]: I actually firstly did as you suggested:-) The reason to put it later is that between replica servers, we will use this optimization a lot. For client requests, we may not use this as the client-lib may run inside 3rd party apps. So it really depends. As long as we have s-local-hash in replica server non-zero, the first condition also applies for the 2nd scenario.

File: src/core/core/service_engine.cpp (hunk 1 above)
Comment (imzhenyu) [inline src/core/core/service_engine.cpp]: ya, right, my bad.

**Pull request merged**